File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int n = s.size();
        stack<int> openBrackets;
        stack<int> stars;

        for (int i = 0; i < n; i++) {
            if (s[i] == '(') {
                openBrackets.push(i);
            } else if (s[i] == '*') {
                stars.push(i);
            } else {
                if (!openBrackets.empty()) {
                    openBrackets.pop();
                } else if (!stars.empty()) {
                    stars.pop();
                } else {
                    return false;
                }
            }
    
        }
        while (!openBrackets.empty() && !stars.empty()) {
            if (openBrackets.top() > stars.top()) {
                return false;
            }
            openBrackets.pop();
            stars.pop();
        }
        return openBrackets.empty();
    }
};

File: Data Structures & Algorithms/validate-parentheses/submission-2.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        umap = {}
        umap[']'] = '['
        umap[')'] = '('
        umap['}'] = '{'

        for c in s:
            if c in umap:
                if stack and stack[-1] == umap[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)
        return len(stack) == 0
